subsys/net/lib/app/net_app.c

```diff
@@ -0,0 +1,1256 @@
+
+#if defined(MBEDTLS_DEBUG_C) && defined(CONFIG_NET_DEBUG_APP)
+	mbedtls_debug_set_threshold(DEBUG_THRESHOLD);
+	mbedtls_ssl_conf_dbg(&ctx->tls.mbedtls.conf, my_debug, NULL);
+#endif
+
+	/* Seed the RNG */
+	mbedtls_entropy_add_source(&ctx->tls.mbedtls.entropy,
+				   ctx->tls.mbedtls.entropy_src_cb,
+				   NULL,
+				   MBEDTLS_ENTROPY_MAX_GATHER,
+				   MBEDTLS_ENTROPY_SOURCE_STRONG);
+
+	ret = mbedtls_ctr_drbg_seed(
+		&ctx->tls.mbedtls.ctr_drbg,
+		mbedtls_entropy_func,
+		&ctx->tls.mbedtls.entropy,
+		(const unsigned char *)ctx->tls.mbedtls.personalization_data,
+		ctx->tls.mbedtls.personalization_data_len);
+	if (ret != 0) {
+		_net_app_print_error("mbedtls_ctr_drbg_seed returned -0x%x",
+				     ret);
+		goto exit;
+	}
+
+	/* Setup SSL defaults etc. */
+	ret = mbedtls_ssl_config_defaults(&ctx->tls.mbedtls.conf,
+					  client_or_server,
+					  MBEDTLS_SSL_TRANSPORT_STREAM,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (ret != 0) {
+		_net_app_print_error("mbedtls_ssl_config_defaults "
+				     "returned -0x%x", ret);
+		goto exit;
+	}
+
+	mbedtls_ssl_conf_rng(&ctx->tls.mbedtls.conf,
+			     mbedtls_ctr_drbg_random,
+			     &ctx->tls.mbedtls.ctr_drbg);
+
+	if (client_or_server == MBEDTLS_SSL_IS_SERVER) {
+		/* Load the certificates and private RSA key. This needs to be
+		 * done by the user so we call a callback that user must have
+		 * provided.
+		 */
+#if defined(CONFIG_NET_APP_SERVER)
+		ret = ctx->tls.mbedtls.cert_cb(ctx, &ctx->tls.mbedtls.srvcert,
+					       &ctx->tls.mbedtls.pkey);
+		if (ret != 0) {
+			goto exit;
+		}
+#endif
+	} else {
+#if defined(CONFIG_NET_APP_CLIENT)
+		ret = ctx->tls.mbedtls.ca_cert_cb(ctx,
+						  &ctx->tls.mbedtls.ca_cert);
+		if (ret != 0) {
+			goto exit;
+		}
+#endif
+	}
+
+#if defined(MBEDTLS_X509_CRT_PARSE_C) && defined(CONFIG_NET_APP_SERVER)
+	if (client_or_server == MBEDTLS_SSL_IS_SERVER) {
+		mbedtls_ssl_conf_ca_chain(&ctx->tls.mbedtls.conf,
+					  ctx->tls.mbedtls.srvcert.next,
+					  NULL);
+
+		ret = mbedtls_ssl_conf_own_cert(&ctx->tls.mbedtls.conf,
+						&ctx->tls.mbedtls.srvcert,
+						&ctx->tls.mbedtls.pkey);
+		if (ret != 0) {
+			_net_app_print_error("mbedtls_ssl_conf_own_cert "
+					     "returned -0x%x", ret);
+			goto exit;
+		}
+	}
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+
+	ret = mbedtls_ssl_setup(&ctx->tls.mbedtls.ssl,
+				&ctx->tls.mbedtls.conf);
+	if (ret != 0) {
+		_net_app_print_error("mbedtls_ssl_setup returned -0x%x", ret);
+		goto exit;
+	}
+
+#if defined(MBEDTLS_X509_CRT_PARSE_C) && defined(CONFIG_NET_APP_CLIENT)
+	if (client_or_server == MBEDTLS_SSL_IS_CLIENT &&
+	    ctx->tls.cert_host) {
+		ret = mbedtls_ssl_set_hostname(&ctx->tls.mbedtls.ssl,
+					       ctx->tls.cert_host);
+		if (ret != 0) {
+			_net_app_print_error(
+				"mbedtls_ssl_set_hostname returned -0x%x",
+				ret);
+			goto exit;
+		}
+	}
+#endif
+
+	NET_DBG("SSL %s setup done",
+		client_or_server == MBEDTLS_SSL_IS_CLIENT ? "client" :
+							    "server");
+
+exit:
```

Comment:
We probably need to call mbedtls_ssl_free(), mbedtls_ssl_config_free(), mbedtls_ctr_drbg_free(), and mbedtls_entropy_free() if something goes wrong.